src/main/php/xp/web/srv/HttpProtocol.class.php

```diff
@@ -153,17 +155,30 @@ public function handleData($socket) {
         clearstatcache();
         \xp::gc();
       }
-    } else if (Input::CLOSE === $input->kind) {
-      $socket->close();
+      return;
+    }
+
+    // Handle request errors and close the socket
+    if (Input::CLOSE === $input->kind) {
+      goto close;
+    } else if (Input::TIMEOUT === $input->kind) {
+      $status= '408 Request Timeout';
+      $error= 'Client failed sending status line and request headers in a timely manner';
+    } else if (Input::EXCESSIVE === $input->kind) {
+      $status= '431 Request Header Fields Too Large';
+      $error= 'Client sent excessively long status line or request headers';
     } else {
-      $error= 'Incomplete HTTP request: "'.addcslashes($input->kind, "\0..\37!\177..\377").'"';
-      $socket->write(sprintf(
-        "HTTP/1.1 400 Bad Request\r\nContent-Type: text/plain\r\nContent-Length: %d\r\nConnection: close\r\n\r\n%s",
-        strlen($error),
-        $error
-      ));
-      $socket->close();
+      $status= '400 Bad Request';
+      $error= 'Client sent incomplete HTTP request: "'.addcslashes($input->buffer, "\0..\37!\177..\377").'"';
     }
+
+    $socket->write(sprintf(
```

Comment:
Logging for this would be nice - however, we currently require a "functioning" `Request` object for this.